homeassistant/components/thread/dataset_store.py

```diff
@@ -310,6 +324,69 @@ def preferred_dataset(self, dataset_id: str) -> None:
         self._preferred_dataset = dataset_id
         self.async_schedule_save()
 
+    async def _set_preferred_dataset_if_only_network(
+        self, dataset_id: str, border_agent_id: str
+    ) -> None:
+        """Set the preferred dataset, unless there are other routers present."""
+        _LOGGER.debug(
+            "_set_preferred_dataset_if_only_network called for router %s",
+            border_agent_id,
+        )
+
+        own_router_evt = Event()
+        other_router_evt = Event()
+
+        @callback
+        def router_discovered(
+            key: str, data: discovery.ThreadRouterDiscoveryData
+        ) -> None:
+            """Handle router discovered."""
+            if not data.border_agent_id:
+                return
+
+            if data.border_agent_id == border_agent_id:
+                own_router_evt.set()
+                return
+
+            other_router_evt.set()
+
+        # Start Thread router discovery
+        thread_discovery = discovery.ThreadRouterDiscovery(
+            self.hass, router_discovered, lambda key: None
+        )
+        await thread_discovery.async_start()
+
+        found_own_router = self.hass.async_create_task(own_router_evt.wait())
+        found_other_router = self.hass.async_create_task(other_router_evt.wait())
+        pending = {found_own_router, found_other_router}
+        while pending:
+            (done, pending) = await wait(
+                pending,
+                timeout=30,
+                return_when=FIRST_COMPLETED,
+            )
+            if found_other_router in done:
+                # We found another router on the network, don't set the dataset
+                # as preferred
+                _LOGGER.debug("Other router found, do not set dataset as default")
+                break
+
+            if not done:
+                if found_own_router in pending:
+                    # Either the router is not there, or mDNS is not working. In any case,
+                    # don't set the router as preferred
+                    _LOGGER.debug("Own router not found, do not set dataset as default")
+                    break
+                # We did not find any other router on the network, mark the dataset as
+                # preferred
+                _LOGGER.debug("No other router found, set dataset as default")
+                self.preferred_dataset = dataset_id
+                break
```

Comment:
This is in the `not done` if, is this correct? It feels like the case where `found_own_router` is done and `found_other_router`  is not is not handled well :thinking:

Reply:
The first check is for `found_other_router` in the `done`-set, if that happens we found another router and break out of the `while` loop.

After that we check for an empty `done`-set, meaning the wait timed out:
- If our own router is still in the `pending`-set, we break out of the `while` loop
- Otherwise, our own router is *not* in the pending set meaning we've found it. Since we've already checked if the other router is in the `done`-set we don't need to check it again.

Reply:
I removed the while loop, the code should now be easier to follow